map/bookmark_manager.cpp

```diff
@@ -404,6 +406,52 @@ void BookmarkManager::ResetRecentlyDeletedBookmark()
   m_recentlyDeletedBookmark.reset();
 }
 
+BookmarkManager::KMLDataCollectionPtr BookmarkManager::GetRecentlyDeletedCategories()
+{
+  auto collection = LoadBookmarks(GetBookmarksDirectory(), kDeletedExtension, KmlFileType::Text,
+                                  [](kml::FileData const &)
+                                  {
+    return true;
+  });
+  return collection;
+}
+
+bool BookmarkManager::IsRecentlyDeletedCategory(const std::string &filePath) const
+{
+  return base::GetFileExtension(filePath) == kDeletedExtension;
+}
+
+void BookmarkManager::RecoverRecentlyDeletedCategoriesAtPaths(std::vector<std::string> const & deletedFilePaths)
+{
+  CHECK_THREAD_CHECKER(m_threadChecker, ());
+  GetPlatform().RunTask(Platform::Thread::File, [this, deletedFilePaths]()
+  {
+    for (auto const & deletedFilePath : deletedFilePaths)
+    {
+      CHECK(IsRecentlyDeletedCategory(deletedFilePath), ("The category at path", deletedFilePath, " should be 'deleted'."));
+      ASSERT(Platform::IsFileExistsByFullPath(deletedFilePath), ("Deleted file should exist to be recovered.", deletedFilePath));
+      auto filePath = base::FilenameWithoutExt(deletedFilePath);
+      if (Platform::IsFileExistsByFullPath(filePath))
+        filePath = GenerateValidAndUniqueFilePathForKML(base::GetNameFromFullPathWithoutExt(deletedFilePath));
+      UpdateLastModifiedTime(deletedFilePath);
```

Comment:
A detailed comment is needed on why is this call required. Is there any way to avoid using it?

Reply:
We should track the `delete` / `recover` action.

Situation:
device 1:  
 - user deletes the category (file is marked by the .deleted ext)
 - could is synced
device 2:
 - user disables the cloud
 - user recovers the category (remove the .deleted ext)
 - user enables the could

Expectation: the file should be recovered.

But we have:
on cloud:
file.kml.deleted (`lastModified` 0001)
on device 1:
file.kml (`lastModified` 0001)

how to decide what is the latest file version?

Updating `lastModified` allows to keep track the file's lifecycle.
The implementation without using `lastModified` does not allow to track it.

The way to avoid this: create the versionig system or some kind of transactions journal.